src/main/java/uk/gov/hmcts/ccd/domain/service/callbacks/CallbackService.java

```diff
@@ -185,13 +152,46 @@ private void sleep(final Integer timeout) {
         }
         final HttpEntity requestEntity = new HttpEntity(callbackRequest, httpHeaders);
 
-        StopWatch sw = new StopWatch();
-        sw.start();
-        ResponseEntity<T> exchange = restTemplate.exchange(url, HttpMethod.POST, requestEntity, clazz);
-        sw.stop();
-        LOG.info("CallbackExecutionTime={} caseType={} event={} url={}", sw.getTotalTimeMillis(), caseTypeId, eventId, url);
-        return ofNullable(exchange);
+        Future<ResponseEntity> future = executorService.submit(() -> restTemplate.exchange(url, HttpMethod.POST, requestEntity, clazz));
+
+        return handleResponse(url, timeout, future, caseTypeId, eventId);
+    }
+
+    @SuppressWarnings("javasecurity:S5145")
+    private <T> Optional<ResponseEntity<T>> handleResponse(final String url, final Integer timeout, final Future<ResponseEntity> future,
+                                                           final String caseTypeId, final String eventId) {
+        try {
```

Comment:
I have some concerns here. Have we considered using the retry and timeout features of Apache HTTP Client library? Rather than implementing a custom retry and timeout mechanism ourselves? From a (very) quick look I've seen that Apache HTTP Client allows setting timeouts per request. And has the ability to configure retries. Perhaps we could find a way to make it work?

Reply:
That was tried before and it failed in DEMO. Besides we thought (Fatih and I) that having the timeout defined encompassing the entire request is actually what services and anyone configuring the timeouts in the spreadsheet would expect. Defining it on http client level does not guarantee breaching the timeout would drop the request as explained before due to the read timeout being socket timeout.

See also:
https://stackoverflow.com/questions/9873810/using-apache-httpclient-how-to-set-the-timeout-on-a-request-and-response